### PR TITLE
Adds initial style guide.

### PR DESCRIPTION
**Pull request by coollog**
Adds an initial style guide as rules that extend the Google Java style guide for our codebase. We will add more rules/amend existing rules as necessary.

Preview at: https://github.com/GoogleContainerTools/jib/blob/style-guide/STYLE_GUIDE.md

### CODE REVIEW

File: STYLE_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Style guide
+
+This style guide and defines specific coding standards and advice for this Java codebase. The rules here are extensions to the [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html).
```
Comment (TadCordle) [inline STYLE_GUIDE.md]: "This style guide ~~and~~ defines"

File: STYLE_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+Please see the [contributing guide](CONTRIBUTING.md) for general guidance for contributing to this project.
+
+### Automatic formatting
+
+Automatic formatting should be performed with `./gradlew goJF` or `./mvnw fmt:format`.
```
Comment (TadCordle) [inline STYLE_GUIDE.md]: Should we mention `build.sh` so people can format all the modules at once?

File: STYLE_GUIDE.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+### Class member order
+
+*Extends [3.4.2](https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents)*
+
+Class members should in the following order in decreasing priority:
```
Comment (TadCordle) [inline STYLE_GUIDE.md]: Missing a word ("Class members should **appear/be?** in the following order")

File: STYLE_GUIDE.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+User-facing methods (such as those in Jib Core) should not have types in their signature that are not standard JDK classes. For example, a parameter should take type `List` rather than Guava's `ImmutableList`.
+
+### Package hierarchy
+
+Packages should depend on each other without cycles.
+
+The following is a list of current `jib-core` packages (under `com.google.cloud.tools.jib`) and their immediate dependencies. These can be amended as code changes, but there should not be cyclical dependencies.
+
+- `api`
```
Comment (TadCordle) [inline STYLE_GUIDE.md]: This section is a bit hard to read, I wonder if we could somehow arrange this in a tree/pyramid structure to make the relationships more clear? Maybe not possible until we fix the cyclical dependencies.

Comment (coollog) [reply STYLE_GUIDE.md]: Yea, I think we can iterate on this to make it more readable as we fix the cycles.

**Comment by TadCordle** [comment]
LGTM for now, will let someone else approve.

File: STYLE_GUIDE.md (hunk 4 above)
Comment (briandealwis) [reply STYLE_GUIDE.md]: Could you use `dot` or something to express this as a graph?

Comment (briandealwis) [reply STYLE_GUIDE.md]: It's a graph generator from [GraphViz](https://www.graphviz.org).  There's a [web generator](http://www.webgraphviz.com).

Comment (coollog) [reply STYLE_GUIDE.md]: Oh cool, I'll try that out.

Comment (coollog) [reply STYLE_GUIDE.md]: Oh man I tried it out but the graph is quite messy:

```
digraph G {
"api"
"async"
"blob" -> "filesystem"
"blob" -> "hash"
"blob" -> "image" [label = "cycle - should fix" ]
"builder" -> "async"
"builder" -> "blob"
"builder" -> "cache"
"builder" -> "configuration"
"builder" -> "docker"
"builder" -> "event"
"builder" -> "filesystem"
"builder" -> "global"
"builder" -> "http"
"builder" -> "image"
"builder" -> "json"
"builder" -> "registry"
"cache" -> "blob"
"cache" -> "filesystem"
"cache" -> "hash"
"cache" -> "image"
"cache" -> "json"
"configuration" -> "cache"
"configuration" -> "filesystem"
"configuration" -> "event"
"configuration" -> "image"
"configuration" -> "registry"
"docker" -> "blob"
"docker" -> "cache"
"docker" -> "image"
"docker" -> "json"
"docker" -> "tar"
"event"
"filesystem"
"frontend" -> "configuration"
"frontend" -> "event"
"frontend" -> "filesystem"
"frontend" -> "image"
"frontend" -> "registry"
"global"
"hash" -> "blob"
"hash" -> "image"
"http" -> "blob"
"image" -> "blob"
"image" -> "configuration" [ label = "cycle - should fix - ImageToJsonTranslator" ]
"image" -> "filesystem"
"image" -> "json"
"image" -> "tar"
"json" -> "blob"
"registry" -> "blob"
"registry" -> "builder" [ label = "cycle - should fix - RegistryClient" ]
"registry" -> "configuration" [ label = "cycle - should fix - DockerConfigCredentialRetriever" ]
"registry" -> "event"
"registry" -> "global"
"registry" -> "http"
"registry" -> "image"
"registry" -> "json"
"tar" -> "blob"
}
```
resulting graph:
graph

File: STYLE_GUIDE.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+Class members should be in the following order, in decreasing priority:
+
+1. Static before non-static
+1. Nested classes/interfaces before fields before constructors before methods
+1. Public before private
+1. Final before non-final
+
+### Public APIs
+
+User-facing methods (such as those in Jib Core) should not have types in their signature that are not standard JDK classes. For example, a parameter should take type `List` rather than Guava's `ImmutableList`.
```
Comment (patflynn) [inline STYLE_GUIDE.md]: presumably you could have jib core types.

File: STYLE_GUIDE.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+### Package hierarchy
+
+Packages should depend on each other without cycles.
```
Comment (patflynn) [inline STYLE_GUIDE.md]: you mean java packages? what about sub-packages? This is a little bit more stringent than what good build systems enforce which is no-cycles between build level modules.

File: STYLE_GUIDE.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+Automatic formatting should be performed with `./gradlew goJF` or `./mvnw fmt:format`. Formatting all projects can be done with `./build.sh format`.
+
+### Class member order
+
+*Extends [3.4.2](https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents)*
+
+Class members should be in the following order, in decreasing priority:
+
+1. Static before non-static
```
Comment (patflynn) [inline STYLE_GUIDE.md]: While I generally agree with tending to order things in more or less this way, I've found the more relaxed logical ordering requirement in the linked google style guide to work pretty well.  In a way, this section goes against the advice in the Google style-guide.

Defining symbols close to first use is a common practice at Google that often helps readability but would be disallowed by this strict ordering.

**Pull request merged**